<a href="https://colab.research.google.com/github/DataRM-BR/gpt-finetune/blob/main/OpenAI_Finetuning_Distill_GPT_4_to_GPT_3_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Fine Tuning GPT-3.5-Turbo

In this notebook, we walk through an example of fine-tuning gpt-3.5-turbo.

Specifically, we attempt to distill GPT-4's knowledge, by generating training data with GPT-4 to then fine-tune GPT-3.5.

All training data is generated using two different sections of our index data, creating both a training and evalution set.

Evaluation is done using the `ragas` library, which we will detail later on.

In [3]:
%pip install llama-index pypdf sentence-transformers ragas

Note: you may need to restart the kernel to use updated packages.


In [4]:
import os
import openai

os.environ["OPENAI_API_KEY"] = "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"
openai.api_key = os.environ["OPENAI_API_KEY"]

## Data Setup

Here, we first down load the PDF that we will use to generate training data.

In [ ]:
!curl https://www.ipcc.ch/report/ar6/wg2/downloads/report/IPCC_AR6_WGII_Chapter03.pdf --output IPCC_AR6_WGII_Chapter03.pdf

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 20.7M  100 20.7M    0     0  16.7M      0  0:00:01  0:00:01 --:--:-- 16.8M


The next step is generating a training and eval dataset.

We will generate 40 questions on different sections of the PDF we downloaded.

We can use GPT-3.5 on the eval questions to get our baseline performance.

Then, we will use GPT-4 on the train questions to generate our training data. The training data will be collected with out `OpenAIFineTuningHandler`.

This step is entirely optional if you don't want to spend the time/tokens -- the eval and training questions are also provided in this folder, as well as the training data!

### Train Generation

In [ ]:
import nltk
import ssl

try:
    _create_unverified_https_context = ssl._create_unverified_context
except AttributeError:
    pass
else:
    ssl._create_default_https_context = _create_unverified_https_context

nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


: 

: 

In [6]:
from llama_index import SimpleDirectoryReader, ServiceContext
from llama_index.llms import OpenAI
from llama_index.evaluation import DatasetGenerator

documents = SimpleDirectoryReader(
    input_files=["Solicitation_RFP.pdf"]
).load_data()

# Shuffle the documents
import random

random.seed(42)
random.shuffle(documents)

gpt_35_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-3.5-turbo", temperature=0.3)
)

In [9]:
question_gen_query = (
    "You are an AI Assistant that reads Requests for Proposals (RFPs). Your task is to "
    "answer spcific client questions. Using the provided context from an "
    "RFP for a Human Services Case Managemet System, formulate "
    "a single question that captures an important piece of information from the "
    "context. Restrict the question to the context information provided."
)

dataset_generator = DatasetGenerator.from_documents(
    documents[:50],
    question_gen_query=question_gen_query,
    service_context=gpt_35_context,
)

In [10]:
print(len(documents))

143


In [11]:
# NOTE: this may take some time. Go grab a coffee!
questions = dataset_generator.generate_questions_from_nodes(num=40)
print("Generated ", len(questions), " questions")

Generated  40  questions


In [12]:
with open("train_questions.txt", "w") as f:
    for question in questions:
        f.write(question + "\n")

### Eval Generation

Now, lets generate questions on a completely different set of documents, in order to create our eval dataset.

In [13]:
dataset_generator = DatasetGenerator.from_documents(
    documents[
        50:
    ],  # since we generated ~1 question for 40 documents, we can skip the first 40
    question_gen_query=question_gen_query,
    service_context=gpt_35_context,
)

In [14]:
# NOTE: this may take some time. Go grab a coffee!
questions = dataset_generator.generate_questions_from_nodes(num=40)
print("Generated ", len(questions), " questions")

Generated  40  questions


In [15]:
with open("eval_questions.txt", "w") as f:
    for question in questions:
        f.write(question + "\n")

## Initial Eval with GPT-3.5-Turbo Query Engine

For this eval, we will be using the [`ragas` evaluation library](https://github.com/explodinggradients/ragas).

Ragas has a ton of evaluation metrics for RAG pipelines, and you can read about them [here](https://github.com/explodinggradients/ragas/blob/main/docs/metrics.md).

For this notebook, we will be using the following two metrics

- `answer_relevancy` - This measures how relevant is the generated answer to the prompt. If the generated answer is incomplete or contains redundant information the score will be low. This is quantified by working out the chance of an LLM generating the given question using the generated answer. Values range (0,1), higher the better.
- `faithfulness` - This measures the factual consistency of the generated answer against the given context. This is done using a multi step paradigm that includes creation of statements from the generated answer followed by verifying each of these statements against the context. The answer is scaled to (0,1) range. Higher the better.

In [16]:
questions = []
with open("eval_questions.txt", "r") as f:
    for line in f:
        questions.append(line.strip())

In [17]:
from llama_index import VectorStoreIndex

# limit the context window to 2048 tokens so that refine is used
gpt_35_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-3.5-turbo", temperature=0.3), context_window=2048
)

index = VectorStoreIndex.from_documents(documents, service_context=gpt_35_context)

query_engine = index.as_query_engine(similarity_top_k=2)

In [18]:
contexts = []
answers = []

for question in questions:
    response = query_engine.query(question)
    contexts.append([x.node.get_content() for x in response.source_nodes])
    answers.append(str(response))

In [19]:
print(contexts)
print(answers)

[['Page 8 \nSolicitation Number RFP-23-CSSD -75                           02/19 rev. of Events. Offers and other information received in response to the solicitation will be \nshown only to authorized City personnel having a legitimate interest in them or persons \nassisting the City in the  evaluation. Offers are not available for public inspection until after \nthe City has posted the award recommendation on the City’s website.  \n \nPRE-AWARD QUALIFICATIONS :  \n1.1 Offeror  must have been in operation a minimum of 5 years. The Offeror ’s \nnormal business activity during the past  5 years will have been for providing \nthe services, or substantially similar services, to those  requested in this \nsolicitation. (This information must be provided in The Submittal section. \nYears in Business and Customer Reference Listing of this solicitation.)  \n1.2 Upon notification of an award the Offeror  will have 10 business days to \nsubmit a complete certificate of insurance in the min imum 

In [21]:
from datasets import Dataset
from ragas import evaluate
from ragas.metrics import answer_relevancy, faithfulness

ds = Dataset.from_dict(
    {
        "question": questions,
        "answer": answers,
        "contexts": contexts,
    }
)

result = evaluate(ds, [answer_relevancy, faithfulness])
print(result)

evaluating with [answer_relevancy]


 33%|███▎      | 1/3 [00:23<00:46, 23.24s/it]Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised APIError: Internal error {
    "error": {
        "message": "Internal error",
        "type": "internal_error",
        "param": null,
        "code": "internal_error"
    }
}
 500 {'error': {'message': 'Internal error', 'type': 'internal_error', 'param': None, 'code': 'internal_error'}} {'Date': 'Thu, 07 Sep 2023 22:01:55 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Content-Length': '152', 'Connection': 'keep-alive', 'vary': 'Origin', 'x-ratelimit-limit-requests': '3500', 'x-ratelimit-limit-tokens': '180000', 'x-ratelimit-remaining-requests': '3499', 'x-ratelimit-remaining-tokens': '179481', 'x-ratelimit-reset-requests': '17ms', 'x-ratelimit-reset-tokens': '173ms', 'x-request-id': 'f8ee059e53d5a3d1f576b014304c9159', 'strict-transport-security': 'max-age=15724800; includeSubDomains', 'CF-Cache-Statu

evaluating with [faithfulness]


  0%|          | 0/3 [00:00<?, ?it/s]Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised APIError: Internal error {
    "error": {
        "message": "Internal error",
        "type": "internal_error",
        "param": null,
        "code": "internal_error"
    }
}
 500 {'error': {'message': 'Internal error', 'type': 'internal_error', 'param': None, 'code': 'internal_error'}} {'Date': 'Thu, 07 Sep 2023 22:06:06 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Content-Length': '152', 'Connection': 'keep-alive', 'vary': 'Origin', 'x-ratelimit-limit-requests': '3500', 'x-ratelimit-limit-tokens': '180000', 'x-ratelimit-remaining-requests': '3499', 'x-ratelimit-remaining-tokens': '179577', 'x-ratelimit-reset-requests': '17ms', 'x-ratelimit-reset-tokens': '141ms', 'x-request-id': '2bf9d6623ac3736ef7b07970cf3734ee', 'strict-transport-security': 'max-age=15724800; includeSubDomains', 'CF-Cache-Status': 'DYN

{'ragas_score': 0.9209, 'answer_relevancy': 0.9785, 'faithfulness': 0.8697}


## GPT-4 to Collect Training Data

Here, we use GPT-4 and the `OpenAIFineTuningHandler` to collect data that we want to train on.

In [22]:
from llama_index import ServiceContext
from llama_index.llms import OpenAI
from llama_index.callbacks import OpenAIFineTuningHandler
from llama_index.callbacks import CallbackManager

finetuning_handler = OpenAIFineTuningHandler()
callback_manager = CallbackManager([finetuning_handler])

gpt_4_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-4", temperature=0.3),
    context_window=2048,  # limit the context window artifically to test refine process
    callback_manager=callback_manager,
)

In [23]:
questions = []
with open("train_questions.txt", "r") as f:
    for line in f:
        questions.append(line.strip())

In [24]:
from llama_index import VectorStoreIndex

index = VectorStoreIndex.from_documents(documents, service_context=gpt_4_context)

query_engine = index.as_query_engine(similarity_top_k=2)

In [25]:
for question in questions:
    response = query_engine.query(question)

## Create Fine-Tuning Data

Fine-Tuning data must be written as a list of messages in a `.jsonl` file. Using the finetuning-handler, we can easily write the messages to a `.jsonl` file.

In [26]:
finetuning_handler.save_finetuning_events("finetuning_events.jsonl")

Wrote 41 examples to finetuning_events.jsonl


## Launch Fine-Tuning Job

In [27]:
%pip install wget

Note: you may need to restart the kernel to use updated packages.


In [28]:
# download launch_training.py and associated scripts
!wget https://raw.githubusercontent.com/jerryjliu/llama_index/main/experimental/openai_fine_tuning/launch_training.py -O launch_training.py
!wget https://github.com/jerryjliu/llama_index/blob/main/experimental/openai_fine_tuning/validate_json.py -O validate_json.py

# [optional] if you want to load the precached events
# !wget https://raw.githubusercontent.com/jerryjliu/llama_index/main/experimental/openai_fine_tuning/finetuning_events.jsonl -O finetuning_events.jsonl

zsh:1: command not found: wget
zsh:1: command not found: wget


In [30]:
# alternative to wget
!curl -O https://raw.githubusercontent.com/jerryjliu/llama_index/main/experimental/openai_fine_tuning/launch_training.py
!curl -O https://raw.githubusercontent.com/jerryjliu/llama_index/main/experimental/openai_fine_tuning/validate_json.py

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1099  100  1099    0     0   4645      0 --:--:-- --:--:-- --:--:--  4757
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  6397  100  6397    0     0  32097      0 --:--:-- --:--:-- --:--:-- 32472


In [39]:
%brew install python

UsageError: Line magic function `%brew` not found.


In [47]:
!python ./launch_training.py ./finetuning_events.jsonl

zsh:1: command not found: python


## Evaluation

After some time, your model will be done training!

The next step is running our fine-tuned model on our eval dataset again to measure any performance increase.

In [ ]:
ft_model_name = "ft:gpt-3.5-turbo-0613:..."

In [ ]:
from llama_index import ServiceContext
from llama_index.llms import OpenAI
from llama_index.callbacks import OpenAIFineTuningHandler
from llama_index.callbacks import CallbackManager


ft_context = ServiceContext.from_defaults(
    llm=OpenAI(model=ft_model_name, temperature=0.3),
    context_window=2048,  # limit the context window artifically to test refine process
)

In [ ]:
questions = []
with open("eval_questions.txt", "r") as f:
    for line in f:
        questions.append(line.strip())

In [ ]:
from llama_index import VectorStoreIndex

index = VectorStoreIndex.from_documents(documents, service_context=ft_context)

query_engine = index.as_query_engine(similarity_top_k=2)

In [ ]:
contexts = []
answers = []

for question in questions:
    response = query_engine.query(question)
    contexts.append([x.node.get_content() for x in response.source_nodes])
    answers.append(str(response))

In [1]:
print(contexts)
print(answers)

NameError: name 'contexts' is not defined

In [ ]:
from datasets import Dataset
from ragas import evaluate
from ragas.metrics import answer_relevancy, faithfulness

ds = Dataset.from_dict(
    {
        "question": questions,
        "answer": answers,
        "contexts": contexts,
    }
)

result = evaluate(ds, [answer_relevancy, faithfulness])
print(result)

evaluating with [answer_relevancy]


100%|██████████| 3/3 [00:50<00:00, 16.92s/it]


evaluating with [faithfulness]


100%|██████████| 3/3 [03:15<00:00, 65.20s/it]


{'ragas_score': 0.8845, 'answer_relevancy': 0.9758, 'faithfulness': 0.8088}


## Exploring Differences

Let's quickly compare the differences in responses, to demonstrate that fine tuning did indeed change something.

In [ ]:
from llama_index import VectorStoreIndex

index = VectorStoreIndex.from_documents(documents)

In [ ]:
questions = []
with open("eval_questions.txt", "r") as f:
    for line in f:
        questions.append(line.strip())

In [ ]:
print(questions[12])

What is a key barrier globally for ocean health, governance, and adaptation to climate change according to the report?


### Original

In [ ]:
from llama_index.response.notebook_utils import display_response
from llama_index import ServiceContext
from llama_index.llms import OpenAI


gpt_35_context = ServiceContext.from_defaults(
    llm=OpenAI(model="gpt-3.5-turbo", temperature=0.3),
    context_window=2048,  # limit the context window artifically to test refine process
)

In [ ]:
query_engine = index.as_query_engine(service_context=gpt_35_context)

response = query_engine.query(questions[12])

display_response(response)

**`Final Response:`** According to the report, a key barrier globally for ocean health, governance, and adaptation to climate change is the availability of technology, knowledge, and financial support, as well as existing governance structures.

### Fine-Tuned

In [ ]:
from llama_index import ServiceContext
from llama_index.llms import OpenAI


ft_context = ServiceContext.from_defaults(
    llm=OpenAI(model=ft_model_name, temperature=0.3),
    context_window=2048,  # limit the context window artifically to test refine process
)

In [ ]:
query_engine = index.as_query_engine(service_context=ft_context)

response = query_engine.query(questions[12])

display_response(response)

**`Final Response:`** The report identifies a broad range of barriers and limits for adaptation to climate change in ecosystems and human systems. These limitations include the availability of technology, knowledge, and financial support, as well as existing governance structures. Existing ocean-governance structures are already facing multi-dimensional, scale-related challenges because of climate change.

As we can see, the fine-tuned model provides a more thorough response! This lines up with the increased faithfullness score from ragas, since the answer is more representative of the retrieved context.

## Conclusion

So, in conclusion, finetuning with only ~61 questions actually helped improve our eval scores!

**answer_relevancy: 0.9778 -> 0.9758**

The answer relenvancy appears to be basically unchanged, between models.

**faithfulness: 0.7638 -> 0.8088**

The faithfulness appears to have been improved! This mains the anwers given better fuffil the original question that was asked.